dpnp/backend/extensions/lapack/heevd_batch.cpp

```diff
@@ -0,0 +1,207 @@
+    std::int64_t w_size = n;
+
+    const std::int64_t lda = std::max<size_t>(1UL, n);
+
+    // Get the number of independent linear streams
+    std::int64_t n_linear_streams =
+        (batch_size > 16) ? 4 : ((batch_size > 4 ? 2 : 1));
+
+    const std::int64_t scratchpad_size =
+        mkl_lapack::heevd_scratchpad_size<T>(exec_q, jobz, upper_lower, n, lda);
+    T *scratchpad = nullptr;
+
+    // Get padding size to ensure memory allocations are aligned to 256 bytes
+    // for better performance
+    std::int64_t padding = 256 / sizeof(T);
+
+    if (scratchpad_size > 0) {
+        // Calculate the total scratchpad memory size needed for all linear
+        // streams with proper alignment
+        size_t alloc_scratch_size =
+            helper::round_up_mult(n_linear_streams * scratchpad_size, padding);
+
+        // Allocate memory for the total scratchpad
+        scratchpad = sycl::malloc_device<T>(alloc_scratch_size, exec_q);
+        if (!scratchpad)
+            throw std::runtime_error("Device allocation for scratchpad failed");
+    }
+
+    // Computation events to manage dependencies for each linear stream
+    std::vector<std::vector<sycl::event>> comp_evs(n_linear_streams, depends);
+
+    std::stringstream error_msg;
+    std::int64_t info = 0;
+
+    // Release GIL to avoid serialization of host task
+    // submissions to the same queue in OneMKL
+    py::gil_scoped_release release;
+
+    for (std::int64_t batch_id = 0; batch_id < batch_size; ++batch_id) {
+        T *a_batch = a + batch_id * a_size;
+        RealT *w_batch = w + batch_id * w_size;
+
+        std::int64_t stream_id = (batch_id % n_linear_streams);
+
+        T *current_scratch_heevd = nullptr;
+        if (scratchpad != nullptr) {
+            current_scratch_heevd = scratchpad + stream_id * scratchpad_size;
+        }
+
+        // Get the event dependencies for the current stream
+        const auto &current_dep = comp_evs[stream_id];
+
+        sycl::event heevd_event;
+        try {
+            heevd_event = mkl_lapack::heevd(
+                exec_q,
+                jobz, // 'jobz == job::vec' means eigenvalues and eigenvectors
+                      // are computed.
+                upper_lower, // 'upper_lower == job::upper' means the upper
+                             // triangular part of A, or the lower triangular
+                             // otherwise
+                n,           // The order of the matrix A (0 <= n)
+                a_batch,     // Pointer to the square A (n x n)
+                             // If 'jobz == job::vec', then on exit it will
+                             // contain the eigenvectors of A
+                lda, // The leading dimension of A, must be at least max(1, n)
+                w_batch, // Pointer to array of size at least n, it will contain
+                         // the eigenvalues of A in ascending order
+                current_scratch_heevd, // Pointer to scratchpad memory to be
```

Comment:
As I can see it's possible to have `current_scratch_heevd == nullptr` while passing here. Is it a valid use case from MKL interface?

Reply:
I have added additional checks for this. Thank you